### ISSUE TITLE
`count_if` functors should not require const call operators

### ISSUE THREAD

**Issue by igordsm** [NVIDIA/thrust#1148]
The following snippet (created from the [example in the docs](https://thrust.github.io/doc/group__counting_gae2f8874093d33f3f0f49b51d8b26438c.html#gae2f8874093d33f3f0f49b51d8b26438c)) does not compile.

```cpp
#include <thrust/count.h>
#include <thrust/device_vector.h>
#include <iostream>

struct is_odd
{
  __host__ __device__
  bool operator()(int &x)
  {
    return x & 1;
  }
};

int main() {
    thrust::device_vector<int> vec(5);
    vec[0] = 0;
    vec[1] = 1;
    vec[2] = 2;
    vec[3] = 3;
    vec[4] = 4;

    int result = thrust::count_if(vec.begin(), vec.end(), is_odd());
    std::cout << result << "\n";

}
```
It seems `is_odd` is lacking the `const` attribute in `operator()`, since it works when I add it. 

###  `nvcc` version
```
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2017 NVIDIA Corporation
Built on Fri_Nov__3_21:07:56_CDT_2017
Cuda compilation tools, release 9.1, V9.1.85
```

**Comment by allisonvacanti** [comment]
Updated title to reflect fix. Discussion in #1088 is relevant.

**Comment by igordsm** [comment]
@allisonvacanti  thank you for the fast response! Should the docs be fixed while #1088 is discussed? I am using `thrust` in a course and multiple students have reported this problem. Unfortunately, none were able to figure out the solution (put `const` in the argument) as the resulting error is not really helpful (for c++ beginners, at least).

**Comment by allisonvacanti** [comment]
Agreed, those issues give exceptionally poor diagnostics and the docs should be updated. I'll look into what's involved in updating those doxygen pages.
